### PR TITLE
Scripts/VioletHold: rewrote the whole instance

### PR DESCRIPTION
**Pull request by joschiwald**
- replace most positions and waypoints with sniffed values
- converted boss scripts in BossAI and TaskScheduler
- added missing emotes
- corrected portal handling and added missing portal locations
- after wipe bosses are replace by placeholder npcs if they are already defeated

thanks @MitchesD for helping

Note: you have to enable mmaps to get it working properly

### CODE REVIEW

File: sql/updates/world/2015_xx_xx_xx_world_violet_hold.sql [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_xevozz' WHERE `entry`=32231; -- Xevozz
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_lavanthor' WHERE `entry`=32237; -- Lavanthor
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_ichoron' WHERE `entry`=32234; -- Ichoron
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_zuramat' WHERE `entry`=32230; -- Zuramat
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_erekem' WHERE `entry`=32226; -- Erekem
+UPDATE `creature_template` SET `minlevel`=76, `maxlevel`=76, `ScriptName`='npc_erekem_guard' WHERE `entry`=32228; -- Erekem Guard
+UPDATE `creature_template` SET `minlevel`=77, `maxlevel`=77, `ScriptName`='boss_moragg' WHERE `entry`=32235; -- Moragg
+
+UPDATE `creature_template` SET `flags_extra`=0 WHERE `entry`=29364;
+-- todo: update immunemasks of dummy boss npcs
```
Comment (ccrs) [inline sql/updates/world/2015_xx_xx_xx_world_violet_hold.sql]: missing todo?

File: src/server/scripts/Northrend/VioletHold/boss_cyanigosa.cpp [hunk 2]
```diff
@@ -42,119 +43,93 @@ enum Yells
...
+                if (!UpdateVictim())
+                    return;
 
-            if (uiBlizzardTimer <= diff)
-            {
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100, true))
-                    DoCast(target, SPELL_BLIZZARD);
-                uiBlizzardTimer = 15000;
-            } else uiBlizzardTimer -= diff;
+                scheduler.Update(diff, [this]
```
Comment (Naios) [inline src/server/scripts/Northrend/VioletHold/boss_cyanigosa.cpp]:
```c++
scheduler.Update(diff, std::bind(&BossAI::DoMeleeAttackIfReady, this));
```

File: src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp [hunk 3]
```diff
@@ -20,80 +20,148 @@
...
-6 - Zuramat
-7 - Cyanigosa */
-
-enum AzureSaboteurSpells
+
+/*
+ * TODO:
+ * - despawn remaining portals after wipe
+ * - replace bosses by dummy npcs also after grid unload
+ */
```
Comment (ccrs) [inline src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp]: another todo?

**Comment by ariel-** [comment]
```diff
---
 .../Northrend/VioletHold/instance_violet_hold.cpp     | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp b/src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp
index b529d09..2905f16 100644
--- a/src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp
+++ b/src/server/scripts/Northrend/VioletHold/instance_violet_hold.cpp
@@ -228,6 +228,12 @@ class instance_violet_hold : public InstanceMapScript
                         if (GetBossState(DATA_ZURAMAT) == DONE)
                             creature->DespawnOrUnsummon();
                         break;
+                    case NPC_VOID_SENTRY_BALL:
+                        if (GetBossState(DATA_ZURAMAT) == DONE)
+                            creature->DespawnOrUnsummon();
+                        else
+                            ZuramatVoidSentries.push_back(creature->GetGUID());
+                        break;
                     default:
                         break;
                 }
@@ -357,7 +363,18 @@ class instance_violet_hold : public InstanceMapScript
                             SetBossState(DATA_2ND_BOSS, state);
 
                         if (state == DONE)
+                        {
                             SetData(DATA_WAVE_COUNT, WaveCount + 1);
+                            if (type == DATA_ZURAMAT)
+                            {
+                                for (GuidVector::const_iterator itr = ZuramatVoidSentries.begin(); itr != ZuramatVoidSentries.end();)
+                                {
+                                    if (Creature* VoidSentry = instance->GetCreature(*itr))
+                                        VoidSentry->DespawnOrUnsummon();
+                                    itr = ZuramatVoidSentries.erase(itr);
+                                }
+                            }
+                        }
                         break;
                     default:
                         break;
@@ -931,6 +948,8 @@ class instance_violet_hold : public InstanceMapScript
             static uint8 const ActivationCrystalCount = 5;
             ObjectGuid ActivationCrystalGUIDs[ActivationCrystalCount];
 
+            GuidVector ZuramatVoidSentries;
+
             uint8 FirstBossId;
             uint8 SecondBossId;
 
-- 
```

Should despawn the Void Sentries (they're another npc_entry and summoned by the Void Sentries that Zuramat summons)

**Comment by ccrs** [comment]
why not add a despawn command on Void Sentries script?

**Comment by ariel-** [comment]
Void Sentries do not have a script of their own, instead use SmartAI, the subspawns (entry 29365) are set to despawn on master (entry = 29364) death, however when the boss dies it only DESPAWNS the master entries, and the subspawns still linger around.

**Comment by ccrs** [comment]
umm then add a setdata on creaturedespawn (29364) before the real summon.Despawn(), and of course change the sai event.

````diff
+void SummonedCreatureDespawn(Creature* summon) override
+{
+    if (summon->GetEntry() == NPC_VOID_SENTRY)
+        summon->AI()->SetData(1, 1);
+   summons.Despawn(summon);
+}
````

**Comment by tkrokli** [comment]
Immense job, thanks for doing all this work, @joschiwald ( & assisting devs). :+1:

**Pull request merged**